Tools/WinMLRunner/src/EventTraceHelper.cpp

```diff
@@ -0,0 +1,559 @@
+///
+/// \param EventRecord  The event record received by EventRecordCallback
+/// \param EventInfo    A struct with event metadata.
+/// \param Result       A string with the formatted data.
+///
+/// \return A DWORD with a windows error value. If the function succeded, it returns
+///     ERROR_SUCCESS.
+///
+DWORD EventTraceHelper::FormatDataForCPUFallback(_In_ const PEVENT_RECORD EventRecord,
+                                               _In_ const PTRACE_EVENT_INFO EventInfo)
+{
+    DWORD status = ERROR_SUCCESS;
+
+    if (EVENT_HEADER_FLAG_32_BIT_HEADER == (EventRecord->EventHeader.Flags & EVENT_HEADER_FLAG_32_BIT_HEADER))
+    {
+        PointerSize = 4;
+    }
+    else
+    {
+        PointerSize = 8;
+    }
+
+    if (EVENT_HEADER_FLAG_STRING_ONLY == (EventRecord->EventHeader.Flags & EVENT_HEADER_FLAG_STRING_ONLY))
+    {
+        // Do nothing
+    }
+    else
+    {
+        PBYTE pUserData = (PBYTE)EventRecord->UserData;
+        PBYTE pEndOfUserData = (PBYTE)EventRecord->UserData + EventRecord->UserDataLength;
+        PWCHAR pPropertyName;
+        wstring operatorType;
+        wstring operatorName;
+        wstring duration;
+        wstring propertyValue;
+        wstring executionProvider;
+        for (USHORT i = 0; i < EventInfo->TopLevelPropertyCount; i++)
+        {
+            pPropertyName = (PWCHAR)((PBYTE)(EventInfo) + EventInfo->EventPropertyInfoArray[i].NameOffset);
+            status = GetFormatedPropertyData(EventRecord, EventInfo, i, pUserData, pEndOfUserData, propertyValue);
+            if (wcscmp(pPropertyName, L"Operator Name") == 0)
+            {
+                operatorType = propertyValue;
+            }
+            else if (wcscmp(pPropertyName, L"Event Name") == 0)
+            {
+                operatorName = TrimOperatorName(propertyValue);
+            }
+            else if (wcscmp(pPropertyName, L"Duration (us)") == 0)
+            {
+                duration = propertyValue;
+            }
+            else if (wcscmp(pPropertyName, L"Execution Provider") == 0)
+            {
+                executionProvider = propertyValue;
+            }
+            if (ERROR_SUCCESS != status)
+            {
+                printf("Failed to format ETW event user data..");
+
+                return status;
+            }
+        }
+        if (commandArgs.UseGPU() && 
+            wcscmp(executionProvider.c_str(), L"CPUExecutionProvider") == 0 && 
+            !operatorName.empty())
+        {
+            wprintf(L"WARNING: CPU fallback detected for operator %s(%s), duration: %s\n", operatorName.c_str(),
+                    operatorType.c_str(), duration.c_str());
+        }
+    }
+
+    return ERROR_SUCCESS;
+}
+
+VOID WINAPI EventTraceHelper::EventRecordCallback(EVENT_RECORD* pEventRecord)
+{
+    // This is where you would get the details of ETW event
+    DWORD status = ERROR_SUCCESS;
+    PTRACE_EVENT_INFO pInfo = NULL;

```

Comment:
can we make this a smart unique_ptr and it should get rid of the goto

Reply:
decided to just restructure the code to get rid of goto, hope it's OK